File: freqStack.java
/*
Intuition

Evidently, we care about the frequency of an element. Let freq be a Map from xx to the number of 
ccurrences of x. Also, we (probably) care about maxfreq, the current maximum frequency of any 
element in the stack. This is clear because we must pop the element with the maximum frequency.
The main question then becomes: among elements with the same (maximum) frequency, 
how do we know which element is most recent? We can use a stack to query this information: 
the top of the stack is the most recent.

To this end, let group be a map from frequency to a stack of elements with that frequency. 
We now have all the required components to implement FreqStack.

Algorithm

Actually, as an implementation level detail, if x has frequency f, 
then we'll have x in all group[i] (i <= f), not just the top. This is because each 
group[i] will store information related to the ith copy of x.

Afterwards, our goal is just to maintain freq, group, and maxfreq as described above.
*/ 

import java.util.*; // TC O(1) // SC O(N)

class xyz {
    Map<Integer, Integer> freq;
    Map<Integer, Stack<Integer>> group;
    int maxfreq;

    public xyz() {
        freq = new HashMap<Integer, Integer>();
        group = new HashMap<Integer, Stack<Integer>>();
        maxfreq = 0;
    }

    public void push(int x) {
        int f = freq.getOrDefault(x, 0) + 1;
        freq.put(x, f);
        if (f > maxfreq)
            maxfreq = f;

        group.computeIfAbsent(f, z-> new Stack<Integer>()).push(x);
    }

    public int pop() {
        int x = group.get(maxfreq).pop();
        freq.put(x, freq.get(x) - 1);
        if (group.get(maxfreq).size() == 0)
            maxfreq--;
        return x;
    }
 
} 

class freqStack {
    public static void main(String [] args){
        xyz obj = new xyz();
        obj.push(5);
        obj.push(7);
        obj.push(5);
        obj.push(7);
        obj.push(4);
        obj.push(5);
        obj.pop();
        obj.pop();
        obj.pop();
        obj.pop();
        System.out.println(obj.freq);
        System.out.println(obj.group);
        
        
    }
}

File: freqStack.py
'''
class freqStack():

	def __init__(self):


	def push(val):


	def pop(self):

'''
import collections
class FreqStack(object): # TC O(1) // SC O(N)
    def __init__(self,obj):
        self.freq = collections.Counter()
        self.group = collections.defaultdict(list)
        self.maxfreq = 0

    def push(self, x):
        f = self.freq[x] + 1
        self.freq[x] = f
        if f > self.maxfreq:
            self.maxfreq = f
        self.group[f].append(x)

    def pop(self):
        x = self.group[self.maxfreq].pop()
        self.freq[x] -= 1
        if not self.group[self.maxfreq]:
            self.maxfreq -= 1

        return x

obj = FreqStack(["FreqStack"])
obj.push(5)
obj.push(7)
obj.push(5)
obj.push(7)
obj.push(4)
obj.push(5)
result=[]
qts=["push","push","push","push","push","push","pop","pop","pop","pop"] 
for qt in qts:
    if qt =="push":
        result.append('null')
    else:
        result.append(obj.pop())
print(result)
print(obj.group)

'''def qt1(op,value):
    for i,j in zip(op,value):
        if i=="push":
            obj.push(j)
        else:
            obj.pop()

print(qt1[["push","push","push","push","push","push","pop","pop","pop","pop"],[5,7,5,7,4,5]])'''










